mender-update/deployments.cpp

```diff
@@ -0,0 +1,207 @@
+#include <common/json.hpp>
+#include <common/log.hpp>
+#include <common/optional.hpp>
+#include <common/path.hpp>
+#include <mender-update/context.hpp>
+
+#include <mender-update/deployments.hpp>
+
+namespace mender {
+namespace update {
+namespace deployments {
+
+using namespace std;
+
+namespace common = mender::common;
+namespace context = mender::update::context;
+namespace error = mender::common::error;
+namespace events = mender::common::events;
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace json = mender::common::json;
+namespace log = mender::common::log;
+namespace optional = mender::common::optional;
+namespace path = mender::common::path;
+
+const DeploymentsErrorCategoryClass DeploymentsErrorCategory;
+
+const char *DeploymentsErrorCategoryClass::name() const noexcept {
+	return "DeploymentsErrorCategory";
+}
+
+string DeploymentsErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case InvalidDataError:
+		return "Invalid data error";
+	case BadResponseError:
+		return "Bad response error";
+	}
+	assert(false);
+	return "Unknown";
+}
+
+error::Error MakeError(DeploymentsErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, DeploymentsErrorCategory), msg);
+}
+
+const string v1_uri = "api/devices/v1/deployments/device/deployments/next";
+const string v2_uri = "api/devices/v2/deployments/device/deployments/next";
+
+error::Error CheckNewDeployments(
+	context::MenderContext &ctx,
+	const string &server_url,
+	http::Client &client,
+	events::EventLoop &loop,
+	APIResponseHandler api_handler) {
+	auto ex_dev_type = ctx.GetDeviceType();
+	if (!ex_dev_type) {
+		return ex_dev_type.error();
+	}
+	string device_type = ex_dev_type.value();
+
+	auto ex_provides = ctx.LoadProvides();
+	if (!ex_provides) {
+		return ex_provides.error();
+	}
+	auto provides = ex_provides.value();
+	if (provides.find("artifact_name") == provides.end()) {
+		return MakeError(InvalidDataError, "Missing artifact name data");
+	}
+	string artifact_name = provides["artifact_name"];
+
+	stringstream ss;
+	ss << R"({"update_control_map": false, "device_provides": {)";
+	ss << R"("device_type": ")";
+	ss << json::EscapeString(device_type);
+	ss << R"(", "artifact_name": ")";
+	ss << json::EscapeString(artifact_name);
+	ss << R"("}})";
+
+	string v2_payload = ss.str();
+	http::BodyGenerator payload_gen = [v2_payload]() {
+		return make_shared<io::StringReader>(v2_payload);
+	};
+
+	// TODO: APIRequest
+	auto v2_req = make_shared<http::OutgoingRequest>();
+	v2_req->SetAddress(path::Join(server_url, v2_uri));
+	v2_req->SetMethod(http::Method::POST);
+	v2_req->SetHeader("Content-Type", "application/json");
+	v2_req->SetHeader("Content-Length", to_string(v2_payload.size()));
+	v2_req->SetHeader("Accept", "application/json");
+	v2_req->SetBodyGenerator(payload_gen);
+
+	string v1_args = "artifact_name=" + http::URLEncode(artifact_name)
```

Comment:
Observation, it'd be nice if `URLEncode` took a dictionary

Reply:
Agree to disagree. It would be nice, but then you have to create the dictionary first and pass it in. And this is not Python/JavaScript :wink:

Reply:
There's nothing keeping it from doing both, no?